engine/common/version/version_control_test.go

```diff
@@ -0,0 +1,345 @@
+		require.NoError(t, err)
+
+		// Create a mock signaler context that expects an error.
+		ictx := irrecoverable.NewMockSignalerContextExpectError(t, ctx, fmt.Errorf(
+			"failed to get highest version beacon for version control: %w",
+			decodeErr))
+
+		// Start the VersionControl component.
+		vc.Start(ictx)
+
+		// Ensure the component is ready before proceeding.
+		unittest.RequireComponentsReadyBefore(t, 2*time.Second, vc)
+
+		// Wait for the asynchronous version controller initialization process.
+		time.Sleep(time.Second)
+	})
+
+	t.Run("start height available", func(t *testing.T) {
+		startHeight := sealedHeader.Height - 5
+		// Set up a mock version beacons storage.
+		versionBeacons := storageMock.NewVersionBeacons(t)
+		// Mock the Highest method to return a version beacon with a specific version.
+		versionBeacons.
+			On("Highest", testifyMock.AnythingOfType("uint64")).
+			Return(func(height uint64) (*flow.SealedVersionBeacon, error) {
+				if height == sealedHeader.Height {
+					return &flow.SealedVersionBeacon{
+						VersionBeacon: unittest.VersionBeaconFixture(
+							unittest.WithBoundaries(
+								flow.VersionBoundary{
+									BlockHeight: startHeight,
+									Version:     "0.0.1",
+								}),
+							unittest.WithBoundaries(
+								flow.VersionBoundary{
+									BlockHeight: sealedHeader.Height,
+									Version:     "0.0.2",
+								}),
+						),
+						SealHeight: sealedHeader.Height,
+					}, nil
+				} else {
+					return nil, nil
+				}
+			}, nil)
+
+		// Create a new VersionControl instance with initial parameters.
+		vc, err := NewVersionControl(unittest.Logger(), versionBeacons, semver.New("0.0.2"), finalizedRootBlockHeight, sealedHeader.Height)
+		require.NoError(t, err)
+
+		// Create a mock signaler context for testing.
+		ictx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+		// Start the VersionControl component.
+		vc.Start(ictx)
+
+		// Ensure the component is ready before proceeding.
+		unittest.RequireComponentsReadyBefore(t, 2*time.Second, vc)
+
+		compatible, err := vc.CompatibleAtBlock(sealedHeader.Height)
+		require.NoError(t, err)
+
+		// Assert that the VersionControl is compatible at the specified block height.
+		assert.True(t, compatible)
+
+		compatible, err = vc.CompatibleAtBlock(startHeight - 1)
+		require.NoError(t, err)
+
+		// Should be incompatible, as lover than start height
+		assert.False(t, compatible)
+	})
+
+	t.Run("start and end heights available", func(t *testing.T) {
+		startHeight := sealedHeader.Height - 5
+
+		endHeight := sealedHeader.Height - 1
+
+		// Set up a mock version beacons storage.
+		versionBeacons := storageMock.NewVersionBeacons(t)
+		// Mock the Highest method to return a version beacon with a specific version.
+		versionBeacons.
+			On("Highest", testifyMock.AnythingOfType("uint64")).
+			Return(func(height uint64) (*flow.SealedVersionBeacon, error) {
+				if height == sealedHeader.Height {
```

Comment:
can you also add a test the covers the case where it has to walk through multiple version beacon events